# Guided Capstone Step 4 - Pre-processing and Training Data Development

### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   **Pre-processing and Training Data Development**

5.  Modeling
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this step of the guided capstone, you'll revisit some of the steps you completed to solve questions in step 4. Revisiting these steps will help you get additional practice completing this kind of work.

**<font color='teal'> Start by loading the necessary packages, as you've done in the previous steps. Print out your current working directory to confirm that you are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3. Remember, it should be saved inside your data subfolder. Print the first five rows.</font>**

In [2]:
df=pd.read_csv('step3_output.csv',index_col=False)
df.head(3)

Unnamed: 0  Unnamed: 0.1                 Name   state  summit_elev  \
0           0             0       Alyeska Resort  Alaska         3939   
1           1             1  Eaglecrest Ski Area  Alaska         2600   
2           2             2     Hilltop Ski Area  Alaska         2090   

   vertical_drop  base_elev  trams  fastEight  fastSixes  ...  \
0           2500        250      1          0          0  ...   
1           1540       1200      0          0          0  ...   
2            294       1796      0          0          0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0             1610.0           113.0             150.0       60.0   
1              640.0            60.0              45.0       44.0   
2               30.0            30.0             150.0       36.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            669.0          65.0          85.0              150.0   
1            350.0          47.0          53.0               90.0   
2             69.0          30.0          34.0              152.0   

   NightSkiing_ac  kmeans  
0           550.0       0  
1             0.0       0  
2            30.0       0  

[3 rows x 29 columns]

In [3]:
df_name=df.Name

In [4]:
df=df.drop(columns=['Name', 'Unnamed: 0', 'Unnamed: 0.1'])
df.head(5)

state  summit_elev  vertical_drop  base_elev  trams  fastEight  \
0   Alaska         3939           2500        250      1          0   
1   Alaska         2600           1540       1200      0          0   
2   Alaska         2090            294       1796      0          0   
3  Arizona        11500           2300       9200      0          0   
4  Arizona        11100           1800       9200      0          0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0          0          2     2       0  ...             1610.0           113.0   
1          0          0     0       0  ...              640.0            60.0   
2          0          0     0       1  ...               30.0            30.0   
3          1          0     2       2  ...              777.0           104.0   
4          0          1     2       3  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0            669.0          65.0          85.0   
1              45.0       44.0            350.0          47.0          53.0   
2             150.0       36.0             69.0          30.0          34.0   
3             122.0       81.0            260.0          89.0          89.0   
4             115.0       49.0            250.0          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  kmeans  
0              150.0           550.0       0  
1               90.0             0.0       0  
2              152.0            30.0       0  
3              122.0             0.0       1  
4              104.0            80.0       1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `State`. Add the dummies back to the dataframe and remove the original column for `State`. </font>**

In [5]:
dfo=df.select_dtypes(include=['object'])  
df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)
df.head(3)

summit_elev  vertical_drop  base_elev  trams  fastEight  fastSixes  \
0         3939           2500        250      1          0          0   
1         2600           1540       1200      0          0          0   
2         2090            294       1796      0          0          0   

   fastQuads  quad  triple  double  ...  state_Rhode Island  \
0          2     2       0       0  ...                   0   
1          0     0       0       4  ...                   0   
2          0     0       1       0  ...                   0   

   state_South Dakota  state_Tennessee  state_Utah  state_Vermont  \
0                   0                0           0              0   
1                   0                0           0              0   
2                   0                0           0              0   

   state_Virginia  state_Washington  state_West Virginia  state_Wisconsin  \
0               0                 0                    0                0   
1               0                 0                    0                0   
2               0                 0                    0                0   

   state_Wyoming  
0              0  
1              0  
2              0  

[3 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing, standardize the scale of the features of the dataframe except the name of the resort, which you don't need in the dataframe for modeling so it can be droppped here as well. You should  hold out your response variable(s) so that you can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as your response for scaling and modeling. Later, you will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave, them in the development dataframe. </font>**

In [6]:
# hold out the response variables for later use
responseV=['AdultWeekend','AdultWeekday','daysOpenLastYear','projectedDaysOpen']
featureV=list(set(df.columns)-set(responseV))
df_response=df[responseV]
df_feature=df[featureV]
df_response.head(100)

AdultWeekend  AdultWeekday  daysOpenLastYear  projectedDaysOpen
0           85.0          65.0        150.000000         150.000000
1           53.0          47.0         45.000000          90.000000
2           34.0          30.0        150.000000         152.000000
3           89.0          89.0        122.000000         122.000000
4           78.0          74.0        115.000000         104.000000
..           ...           ...               ...                ...
95          47.0          39.0        115.103943         120.053004
96          35.0          35.0        115.103943         120.053004
97          69.0          69.0        100.000000         108.000000
98          48.0          38.0         80.000000         120.053004
99          65.0          49.0         95.000000         115.000000

[100 rows x 4 columns]

In [7]:
# all numeric data scaled to 0 mean and unit std
from sklearn import preprocessing
X = df_feature.values
y = df_response['AdultWeekend'].values
X_scale=preprocessing.scale(X)
y_scale=preprocessing.scale(y)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selectionm import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. You will start by using the adult weekend ticket price as your response variable for modeling.</font>**

In [8]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_scale, y_scale, test_size=0.25, random_state=0)

Here, you'll start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5 - Modeling

You've cleaned and prepared the datasets, so now it's time to get into the most exciting part of this process: modeling! In this exercise, you will build three different models and compare each model's performance. In the end, you will choose the best model for proving insights to Big Mountain management.  

1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [9]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import r2_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [10]:
# prediction
y_pred=model.predict(X_test)
# R-squared value
print('r2_score for model 1: ',r2_score(y_test, y_pred))

r2_score for model 1:  -5.139762450407895e+21


In [11]:
# mean_absolute_error
print('mean_absolute_error for model 1: ',mean_absolute_error(y_test, y_pred))

mean_absolute_error for model 1:  7981171317.542733


## Review Model Outcomes — Iterate over additional models as needed.

In [12]:
print(lm.intercept_)

-1927368522.1387022


In [13]:

coeff=lm.coef_
print(coeff)

[ 2.18202499e+07  5.95045416e+06  2.32528923e+07  2.64265853e+07
  2.32528922e+07  3.39673033e+10 -1.18576594e-01  2.02654541e+07
  2.22640373e-02 -2.71517412e+10  1.66504159e+07  5.95045415e+06
  1.18465245e+07  8.40241419e+06  1.85578281e+07  6.12437725e-02
  1.76329889e+07 -3.56790365e+10  2.25503387e+07 -2.13446354e+10
 -2.69279337e+00  2.02654541e+07  8.40241402e+06  8.40241411e+06
  1.32244873e+07 -2.89796311e+10  2.70046689e+07  2.52179140e+07
  3.24778564e+07 -1.38372421e-01  1.18465246e+07  1.25670910e-01
  2.10597623e+07  1.18465246e+07 -8.59102675e+09  4.11014557e-02
 -4.80875969e-02 -5.49397469e-02  1.32244874e+07  1.18031645e+00
  1.02751144e+07  1.18465246e+07  1.85578281e+07  1.02751144e+07
  8.40241403e+06  3.06502774e+07  1.76329886e+07 -2.39271208e+10
  1.44643957e+07 -7.38164873e+09  2.79506969e+00  7.64427823e+10
  2.98053741e-01  8.40241411e+06 -1.72990368e+10  1.94331789e+07]


**<font color='teal'> Print the intercept value from the linear model. </font>**

In [14]:
df_res1=pd.DataFrame()
df_res1['feature']=df_feature.columns
df_res1['values']=abs(lm.coef_)
df_res1.sort_values(by=['values'],ascending=False)

feature        values
51          total_chairs  7.644278e+10
17    state_Rhode Island  3.567904e+10
5              fastEight  3.396730e+10
25             fastQuads  2.897963e+10
9                surface  2.715174e+10
47                double  2.392712e+10
19                triple  2.134464e+10
54                  quad  1.729904e+10
34             fastSixes  8.591027e+09
49                 trams  7.381649e+09
28        state_New York  3.247786e+07
45        state_Michigan  3.065028e+07
26        state_Colorado  2.700467e+07
3       state_California  2.642659e+07
27    state_Pennsylvania  2.521791e+07
2    state_New Hampshire  2.325289e+07
4        state_Wisconsin  2.325289e+07
18         state_Vermont  2.255034e+07
0        state_Minnesota  2.182025e+07
32            state_Utah  2.105976e+07
21           state_Idaho  2.026545e+07
7          state_Montana  2.026545e+07
55   state_Massachusetts  1.943318e+07
42      state_Washington  1.855783e+07
14          state_Oregon  1.855783e+07
16           state_Maine  1.763299e+07
46      state_New Mexico  1.763299e+07
10         state_Wyoming  1.665042e+07
48  state_North Carolina  1.446440e+07
38     state_Connecticut  1.322449e+07
24            state_Ohio  1.322449e+07
33          state_Nevada  1.184652e+07
41        state_Virginia  1.184652e+07
30   state_West Virginia  1.184652e+07
12        state_Illinois  1.184652e+07
43          state_Alaska  1.027511e+07
40            state_Iowa  1.027511e+07
13      state_New Jersey  8.402414e+06
53         state_Indiana  8.402414e+06
23        state_Missouri  8.402414e+06
44         state_Arizona  8.402414e+06
22    state_South Dakota  8.402414e+06
1         state_Maryland  5.950454e+06
11       state_Tennessee  5.950454e+06
50             base_elev  2.795070e+00
20           summit_elev  2.692793e+00
39         vertical_drop  1.180316e+00
52                  Runs  2.980537e-01
29        Snow Making_ac  1.383724e-01
31        NightSkiing_ac  1.256709e-01
6                 kmeans  1.185766e-01
15          TerrainParks  6.124377e-02
37     SkiableTerrain_ac  5.493975e-02
36         LongestRun_mi  4.808760e-02
35             yearsOpen  4.110146e-02
8        averageSnowfall  2.226404e-02

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the `AdultWeekend` resulting value. Also, because you took the time to scale your x values in the training data, you can compare each of the coefficients for the features to determine the feature's importance. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what you are looking for is the magnitude of impact on your response variable.

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that you care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [15]:
# remove 'state' columns
df_feature2=df_feature.drop(df_feature.filter(regex='state').columns, axis=1)
X = df_feature2.values
# scale features
X_scale=preprocessing.scale(X)
#train/test split
X_train, X_test, y_train, y_test = train_test_split(X_scale, y_scale, test_size=0.25, random_state=0)
# model fit
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [16]:
df_feature2.head(2)

fastEight  kmeans  averageSnowfall  surface  TerrainParks  triple  \
0          0       0            669.0        2           2.0       0   
1          0       0            350.0        0           1.0       0   

   summit_elev  fastQuads  Snow Making_ac  NightSkiing_ac  ...  yearsOpen  \
0         3939          2           113.0           550.0  ...       60.0   
1         2600          0            60.0             0.0  ...       44.0   

   LongestRun_mi  SkiableTerrain_ac  vertical_drop  double  trams  base_elev  \
0            1.0             1610.0           2500       0      1        250   
1            2.0              640.0           1540       4      0       1200   

   total_chairs  Runs  quad  
0             7  76.0     2  
1             4  36.0     0  

[2 rows x 21 columns]

In [17]:

# prediction
y_pred=model.predict(X_test)
# R-squared value
print('r2_score for model 2: ',r2_score(y_test, y_pred))

# mean_absolute_error
print('mean_absolute_error for model 2: ',mean_absolute_error(y_test, y_pred))

print(lm.intercept_)
print(lm.coef_)

r2_score for model 2:  0.3802717550589664
mean_absolute_error for model 2:  0.5710571792372419
-0.0030054572912728192
[ 0.         -0.26146903  0.04811203 -0.16258668  0.13837242  0.09783591
 -0.16180843  0.14957888 -0.06758799  0.02690609 -0.06689387  0.0384714
  0.02749107 -0.04479664  0.46188587  0.07251289 -0.08802952  0.30573782
  0.08012686  0.03107319  0.20845578]


In [18]:

df_res2=pd.DataFrame()
df_res2['feature']=df_feature2.columns
df_res2['values']=abs(lm.coef_)
df_res2.sort_values(by=['values'],ascending=False)

feature    values
14      vertical_drop  0.461886
17          base_elev  0.305738
1              kmeans  0.261469
20               quad  0.208456
3             surface  0.162587
6         summit_elev  0.161808
7           fastQuads  0.149579
4        TerrainParks  0.138372
5              triple  0.097836
16              trams  0.088030
18       total_chairs  0.080127
15             double  0.072513
8      Snow Making_ac  0.067588
10          fastSixes  0.066894
2     averageSnowfall  0.048112
13  SkiableTerrain_ac  0.044797
11          yearsOpen  0.038471
19               Runs  0.031073
12      LongestRun_mi  0.027491
9      NightSkiing_ac  0.026906
0           fastEight  0.000000

**<font color='teal'> When reviewing your new model coefficients, note that `summit_elev` is now in the number two spot. From a managerial perspective, this is also difficult to change and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [19]:

# remove 'summit_elev' and 'vertical_drop' columns as managers do not have handle on these
to_drop=['summit_elev','vertical_drop']
df_feature3=df_feature2.drop(to_drop,axis=1)

In [20]:
df_feature3.head(2)

fastEight  kmeans  averageSnowfall  surface  TerrainParks  triple  \
0          0       0            669.0        2           2.0       0   
1          0       0            350.0        0           1.0       0   

   fastQuads  Snow Making_ac  NightSkiing_ac  fastSixes  yearsOpen  \
0          2           113.0           550.0          0       60.0   
1          0            60.0             0.0          0       44.0   

   LongestRun_mi  SkiableTerrain_ac  double  trams  base_elev  total_chairs  \
0            1.0             1610.0       0      1        250             7   
1            2.0              640.0       4      0       1200             4   

   Runs  quad  
0  76.0     2  
1  36.0     0

In [21]:
X = df_feature3.values
# scale features
X_scale=preprocessing.scale(X)
#train/test split
X_train, X_test, y_train, y_test = train_test_split(X_scale, y_scale, test_size=0.25, random_state=0)
# model fit
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [22]:
# prediction
y_pred=model.predict(X_test)
# R-squared value
print('r2_score for model 3: ',r2_score(y_test, y_pred))

# mean_absolute_error
print('mean_absolute_error for model 3: ',mean_absolute_error(y_test, y_pred))

print(lm.intercept_)
print(lm.coef_)

r2_score for model 3:  0.3139719878303303
mean_absolute_error for model 3:  0.6150546329975021
-0.008315753685462082
[ 0.         -0.2350093   0.1087488  -0.18685342  0.15038673  0.09296908
  0.1840914  -0.05979009  0.02479965 -0.08271077  0.03097439  0.17231089
 -0.00662713  0.0832289  -0.06908475  0.20948974  0.08429727  0.15243403
  0.19811064]


In [23]:
df_res3=pd.DataFrame()
df_res3['feature']=df_feature3.columns
df_res3['values']=abs(lm.coef_)
df_res3.sort_values(by=['values'],ascending=False)

feature    values
1              kmeans  0.235009
15          base_elev  0.209490
18               quad  0.198111
3             surface  0.186853
6           fastQuads  0.184091
11      LongestRun_mi  0.172311
17               Runs  0.152434
4        TerrainParks  0.150387
2     averageSnowfall  0.108749
5              triple  0.092969
16       total_chairs  0.084297
13             double  0.083229
9           fastSixes  0.082711
14              trams  0.069085
7      Snow Making_ac  0.059790
10          yearsOpen  0.030974
8      NightSkiing_ac  0.024800
12  SkiableTerrain_ac  0.006627
0           fastEight  0.000000

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell. You will explain your selection during the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -5.3 |7981171317.54  |-|
| Model 2. | 0.38|0.57 |'state'|
| Model 3. | 0.31 |0.61 |'state','summit_elev','base_elev'|

Model Selection: